app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java

```diff
@@ -331,6 +358,127 @@ public void onComplete() { }
         };
     }
 
+    @Override
+    public boolean onOptionsItemSelected(final MenuItem item) {
+        switch (item.getItemId()) {
+            case R.id.menu_item_removeWatched:
+                new AlertDialog.Builder(getActivity())
+                        .setMessage(R.string.remove_watched_popup_warning)
+                        .setTitle(R.string.remove_watched_popup_title)
+                        .setPositiveButton(R.string.yes,
+                                (DialogInterface d, int id) -> removeWatchedStreams(false))
+                        .setNeutralButton(
+                                R.string.remove_watched_popup_yes_and_partially_watched_videos,
+                                (DialogInterface d, int id) -> removeWatchedStreams(true))
+                        .setNegativeButton(R.string.cancel,
+                                (DialogInterface d, int id) -> d.cancel())
+                        .create()
+                        .show();
+                break;
+            default:
+                return super.onOptionsItemSelected(item);
+        }
+        return true;
+    }
+
+    public void removeWatchedStreams(final boolean removePartiallyWatched) {
+        if (removeWatchedDisposable != null && !removeWatchedDisposable.isDisposed()) {
+            // already running
+            return;
+        }
+        showLoading();
+
+        removeWatchedDisposable = playlistManager.getPlaylistStreams(playlistId)
+                .subscribeOn(Schedulers.io())
+                .map((List<PlaylistStreamEntry> playlist) -> {
+                    //Playlist data
+                    Iterator<PlaylistStreamEntry> playlistIter = playlist.iterator();
+
+                    //History data
+                    HistoryRecordManager recordManager = new HistoryRecordManager(getContext());
+                    Iterator<StreamHistoryEntry> historyIter = recordManager
+                            .getStreamHistorySortedById().blockingFirst().iterator();
+                    Iterator<StreamStateEntity> streamStatesIter = null;
+
+                    //Remove Watched, Functionality data
+                    List<PlaylistStreamEntry> notWatchedItems = new ArrayList<>();
+                    boolean thumbnailVideoRemoved = false;
+
+                    if (!removePartiallyWatched) {
+                        streamStatesIter = recordManager.loadLocalStreamStateBatch(playlist)
```

Comment:
This could be moved down I think

Reply:
The space is to separate the variables, and the functions

Reply:
I think he meant to move the
```
streamStatesIter = recordManager.loadLocalStreamStateBatch(playlist)
                                .blockingGet().iterator();
```
block under the `} else {` branch below (where `removePartiallyWatched == false`), i.e. after line 432, so that a check is removed and also less space is used.

Reply:
I think he meant to move the
```
streamStatesIter = recordManager.loadLocalStreamStateBatch(playlist)
                                .blockingGet().iterator();
```
block under the `} else {` branch below (where `removePartiallyWatched == false`), i.e. after line 432, so that a check is removed and also less space is used.

Reply:
Yes, and its initial declaration as well

Reply:
Ah, I'll fix it in the next commit